web/docs/auth/auth-hooks.md

```diff
@@ -0,0 +1,293 @@
+---
+title: Auth Hooks
+---
+
+import { EmailPill, UsernameAndPasswordPill, GithubPill, GooglePill, KeycloakPill } from "./Pills";
+
+
+Sometimes you need to do some action _before_ the user signs up or _after_ they sign up. Or you need to access the OAuth redirect URL or the OAuth access token. For these cases, you can use the Auth Hooks.
+
+## Available Hooks
+
+For every auth hook you want to use, you need to declare it in the `auth` dict in your Wasp file:
+
+```wasp
+app myApp {
+  wasp: {
+    version: "^0.13.0"
+  },
+  auth: {
+    userEntity: User,
+    methods: {
+      ...
+    },
+    onBeforeSignup: import { onBeforeSignup } from "@src/auth/hooks.js",
+    onAfterSignup: import { onAfterSignup } from "@src/auth/hooks.js",
+    onBeforeOAuthRedirect: import { onBeforeOAuthRedirect } from "@src/auth/hooks.js",
+    onAfterOAuthTokenReceived: import { onAfterOAuthTokenReceived } from "@src/auth/hooks.js",
+  },
+}
+```
+
+Wasp _awaits_ the async hook functions in various stages of the auth process. The hooks are called with an `args` object that contains the necessary information for the hook to do its job. 
+
+The return value is ignored for all hooks except `onBeforeOAuthRedirect` which enables you to modify the OAuth redirect URL.
+
+
+### `onBeforeSignup` hook
+
+
+This hook is called before the user signs up. It is an async function that Wasp awaits before proceeding with the signup.
+
+
+:::info
+
+It can be useful if you want to reject a user based on some criteria before they sign up.
+:::
+
+Works with <EmailPill /> <UsernameAndPasswordPill /> <GithubPill /> <GooglePill /> <KeycloakPill />
+
+Its **input** is an `args` object with the following properties:
+
+- `providerId: { providerName: string, providerUserId: string }` - The provider ID of the user
+- `prisma: PrismaClient` - Prisma client instance
+- `req: Request` - Express request object
+- `hookName: string` - The name of the hook that is being called (in this case, it will be `'onBeforeSignup'`)
+
+The **return value** of this hook is ignored.
+
+<Tabs groupId="js-ts">
+<TabItem value="js" label="JavaScript">
+
+```wasp title="main.wasp"
+app myApp {
+  ...
+  auth: {
+    ...
+    onBeforeSignup: import { onBeforeSignup } from "@src/auth/hooks",
+  },
+}
+```
+
+```js title="src/auth/hooks.js"
+export const onBeforeSignup = async ({ providerId, prisma, req, hookName }) => {
+  // Do something before signup
+}
+```
+
+</TabItem>
+<TabItem value="ts" label="TypeScript">
+
+```wasp title="main.wasp"
+app myApp {
+  ...
+  auth: {
+    ...
+    onBeforeSignup: import { onBeforeSignup } from "@src/auth/hooks",
+  },
+}
+```
+
+```ts title="src/auth/hooks.ts"
+import type { OnBeforeSignupHookFn } from 'wasp/server/auth'
+
+export const onBeforeSignup: OnBeforeSignupHookFn = async ({ providerId, prisma, req, hookName }) => {
```

Comment:
TODO: add better example e.g. throw a `HttpError` to show an error message on the client.